src/jvmMain/kotlin/io/github/fourlastor/wilds_launcher/releases/services/GitHubReleaseService.kt

```diff
@@ -0,0 +1,126 @@
+package io.github.fourlastor.wilds_launcher.releases.services
+
+import io.github.fourlastor.wilds_launcher.ui.FILENAME
+import io.github.fourlastor.wilds_launcher.ui.FILENAME_ALT
+import io.github.fourlastor.wilds_launcher.unzipArchive
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import java.io.BufferedInputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.net.HttpURLConnection
+import java.net.URL
+
+const val READ_SIZE_IN_BYTES = 1024
+
+class GitHubReleaseService(
+    private val installDirectory : File
+) : ReleaseService {
+    override fun findInstallation(): File? {
+        return installDirectory.walkTopDown()
+            .filter { it.isFile }
+            .filter { it.name == FILENAME || it.name == FILENAME_ALT }
+            .firstOrNull()
+    }
+
+    override fun getLatestReleaseVersion() : String? {
+        return getLatestReleaseProperty("name")
+    }
+
+    override fun getLatestReleaseChangelog() : String? {
+        return getLatestReleaseProperty("body")
+    }
+
+    override fun getLatestReleaseSizeInBytes(): Long? {
+        val version = getLatestReleaseVersion() ?: return null
+        val connection = getURL(version).openConnection()
+
+        return connection.contentLength.toLong()
+    }
+
+    override fun downloadLatestRelease(onProgressChanged: (Float) -> Unit) : File? {
+        val version = getLatestReleaseVersion()
+
+        if (version == null) {
+            println("Could not get latest version.")
+            return null
+        }
+
+        val connection = getURL(version).openConnection()
+        val totalBytesToDownload = connection.contentLength
+
+        val filename = getFilename()
+        val file = File(filename)
+
+        println("Starting download of latest release.")
+
+        BufferedInputStream(connection.getInputStream()).use { bufferedInputStream ->
+            FileOutputStream(filename).use { fileOutputStream ->
+                val bytes = ByteArray(READ_SIZE_IN_BYTES)
+
+                var downloadedBytes: Int
+                var totalBytesDownloaded = 0
+
+                while (bufferedInputStream.read(bytes, 0, READ_SIZE_IN_BYTES).also { downloadedBytes = it } != -1) {
+                    fileOutputStream.write(bytes, 0, downloadedBytes)
+                    totalBytesDownloaded += downloadedBytes
+
+                    val progress = (totalBytesDownloaded.toFloat() / totalBytesToDownload.toFloat())
+                    println("$totalBytesDownloaded/$totalBytesToDownload (${kotlin.math.floor(progress * 100).toInt()}%) bytes downloaded...")
+
+                    onProgressChanged(progress)
+                }
+            }
+        }
```

Comment:
[OkHttp](https://square.github.io/okhttp/) and [Retrofit](https://square.github.io/retrofit/) would make this very easy, break less than the default java implementation and give you a coroutine api 

For filesystem access I think [Okio's filesystem](https://square.github.io/okio/file_system/) would be also great

(probably worth doing this in another PR tho, as functionally it would be the same)